core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -368,35 +407,44 @@ public void onFailure(Throwable e) {
 
         @Override
         protected void doRun() {
-            if (checkBlocks() == false) {
+            // shardID added to the request, as a side effect of calling resolveRequest
+            if (checkBlocksAndResolveRequest() == false) {
```

Comment:
Is there a good reason for combining these checks? I'm not a fan of the side-effectful request mutating (but I understand why it's done), but at least can it be brought into `doRun`?